plugins/inputs/apcupsd/apcupsd.go

```diff
@@ -0,0 +1,103 @@
+package apcupsd
+
+import (
+	"context"
+	"net/url"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/mdlayher/apcupsd"
+)
+
+const defaultAddress = "tcp://127.0.0.1:3551"
+
+var defaultTimeout = internal.Duration{Duration: time.Duration(time.Second * 5)}
+
+type ApcUpsd struct {
+	Servers []string
+	Timeout internal.Duration
+}
+
+func (_ *ApcUpsd) Description() string {
+	return "Monitor APC UPSes connected to apcupsd"
+}
+
+var sampleConfig = `
+  # a list of running apcupsd server to connect to. 
+  # If not provided will default to tcp://127.0.0.1:3551
+  servers = ["tcp://127.0.0.1:3551"]
+  timeout = "5s"
+`
+
+func (_ *ApcUpsd) SampleConfig() string {
+	return sampleConfig
+}
+
+func (h *ApcUpsd) Gather(acc telegraf.Accumulator) error {
+	for _, addr := range h.Servers {
+		addrBits, err := url.Parse(addr)
+		if err != nil {
+			return err
+		}
+		if addrBits.Scheme == "" {
+			addrBits.Scheme = "tcp"
+		}
+
+		status, err := fetchStatus(addrBits, h.Timeout.Duration)
+		if err != nil {
+			return err
+		}
+
+		tags := map[string]string{
+			"serial":   status.SerialNumber,
+			"ups_name": status.UPSName,
+			"online":   status.Status,
+		}
+
+		fields := map[string]interface{}{
+			"online":                 boolToInt(status.Status == "ONLINE"),
```

Comment:
We should have this be a key that is different from the tag name to simplify querying the data, if we change the tag to `status` then this should be `status_code`.

Reply:
status can be all those things. so the most correct think here would be to put `status_online` as a 1/0 and keep the whole text in tag `status`

```
   // Now output status in human readable form
   statstr = "";
   if (status & UPS_calibration)
      statstr += "CAL ";
   if (status & UPS_trim)
      statstr += "TRIM ";
   if (status & UPS_boost)
      statstr += "BOOST ";
   if (status & UPS_online)
      statstr += "ONLINE ";
   if (status & UPS_onbatt)
      statstr += "ONBATT ";
   if (status & UPS_overload)
      statstr += "OVERLOAD ";
   if (status & UPS_battlow)
      statstr += "LOWBATT ";
   if (status & UPS_replacebatt)
      statstr += "REPLACEBATT ";
   if (!(status & UPS_battpresent))
      statstr += "NOBATT ";
```

Reply:
I see, sets are a little tricky for us to encode nicely.

I think this combined status string might be hard to use in queries, because you may need to look for many string combinations to select what the series you are interested in. Looking at that code snippet and also these [status-bits][], and doing more interpretation (which could get us into trouble), it seems like the UPS status can be consider one of: online, on_battery, or unknown.  The battery can be: okay, low, no_battery, or replace.  Finally there are warning that output is overloaded, and informational messages about cal, trim, and boost.

What if we expose the ups status values and battery status as tags + enum field, and the other status flags as boolean fields.  I'm on the fence if overload should be a tag.

```
apcupsd,status=online,battery=ok status_code=0i,battery_code=0i,overload=false,smarttrim=false,smartboost=false,calibration=false
```

[status-bits]: http://www.apcupsd.org/manual/manual.html#status-bits